src/daft-recordbatch/src/ops/window.rs

```diff
@@ -127,19 +125,114 @@ impl RecordBatch {
+                    total_rows,
+                    agg_state,
+                )
+            }
+            Ok(None) => {
+                // Otherwise, use the non-incremental implementation
+                self.window_agg_rows(
+                    agg_expr,
+                    &name,
+                    dtype,
+                    start_boundary,
+                    end_boundary,
+                    min_periods,
+                    total_rows,
+                )
+            }
+            Err(e) => {
+                // Error during setup (e.g., type incompatibility), propagate it
+                Err(e)
+            }
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn window_agg_rows_incremental<'a>(
+        &'a self,
+        name: &str,
+        start_boundary: Option<i64>,
+        end_boundary: Option<i64>,
+        min_periods: usize,
+        total_rows: usize,
+        mut agg_state: Box<dyn WindowAggStateOps<'a> + 'a>,
+    ) -> DaftResult<Self> {
+        // Track previous window boundaries
+        let mut prev_frame_start = 0;
+        let mut prev_frame_end = 0;
+        let mut validity = MutableBitmap::with_capacity(total_rows); // TODO: probably possible to compute directly
+
+        for row_idx in 0..total_rows {
+            // Calculate frame bounds for this row
+            let frame_start = match start_boundary {
+                None => 0, // Unbounded preceding
+                Some(offset) => row_idx
+                    .saturating_add_signed(offset as isize)
+                    .min(total_rows),
+            };
+
+            let frame_end = match end_boundary {
+                None => total_rows, // Unbounded following
+                Some(offset) => (row_idx + 1)
+                    .saturating_add_signed(offset as isize)
+                    .min(total_rows),
+            };
+
+            let Some(frame_size) = frame_end.checked_sub(frame_start) else {
+                return Err(DaftError::ValueError(
+                    "Negative frame size is not allowed".into(),
+                ));
+            };
+
+            // Check min_periods requirement
+            if frame_size >= min_periods {
+                // Remove values that left the window (values that were in the previous window but not in the current one)
+                if frame_start > prev_frame_start {
+                    agg_state.remove(prev_frame_start, frame_start)?;
+                }
+
+                // Add new values that entered the window (values that are in the current window but weren't in the previous)
+                if frame_end > prev_frame_end {
+                    agg_state.add(prev_frame_end, frame_end)?;
+                }
+
+                // Update previous boundaries for the next iteration
+                prev_frame_start = frame_start;
+                prev_frame_end = frame_end;
+                validity.push(true);
+            } else {
+                validity.push(false);
```

Comment:
Depending on what the default value for MutableBitmap is, one of these isn't necessary